File: src/main/java/edu/neu/coe/info6205/sort/simple/MergeSortBasic.java
package edu.neu.coe.info6205.sort.simple;

import java.util.Arrays;

public class MergeSortBasic<X extends Comparable<X>> implements Sort<X> {

    /**
     * Constructor for InsertionSort
     *
     * @param helper an explicit instance of Helper to be used.
     */
    public MergeSortBasic(Helper<X> helper) {
        this.helper = helper;
    }

    public MergeSortBasic() {
        this(new Helper<>("MergeSort"));
    }

    @Override
    public X[] sort(X[] xs, boolean makeCopy) {
        getHelper().setN(xs.length);
        X[] result = makeCopy ? Arrays.copyOf(xs, xs.length) : xs;
        aux = Arrays.copyOf(xs, xs.length); // TODO don't copy but just allocate
        // TODO make this consistent with other uses of sort where the upper limit of the range is result.length
        sort(result, 0, result.length);
        return result;
    }

    @Override
    public void sort(X[] a, int from, int to) {
        @SuppressWarnings("UnnecessaryLocalVariable") int lo = from;
        int hi = to;
        if (hi <= lo + 1) return;
        int mid = from + (to - from)/2;
        sort(a, lo, mid);
        sort(a, mid, hi);
        System.arraycopy(a, from, aux, from, to - from);
        merge(aux, a, lo, mid, hi);
    }

    private void merge(X[] aux, X[] a, int lo, int mid, int hi) {
        int i = lo;
        int j = mid;
        for (int k = lo; k < hi; k++)
            if (i >= mid) a[k] = aux[j++];
        else if (j >= hi) a[k] = aux[i++];
        else if (helper.less(aux[j], aux[i])) a[k] = aux[j++];
        else a[k] = aux[i++];
    }

    @Override
    public Helper<X> getHelper() {
        return helper;
    }

    private X[] aux = null;
    private final Helper<X> helper;
}

File: Python/src/sort/simple/merge_sort_basic.py
from typing import List

from sort.simple.helper import Helper
from sort.simple.sort import Sort
from util.generic_type import C


class MergeSortBasic(Sort[C]):

    def __init__(self, helper: Helper[C] = None):
        if helper is None:
            helper = Helper[C]("MergeSort")
        self.helper = helper

        self.aux: List[C] = []

    def sort(self, xs: List[C], inplace: bool = False) -> List[C]:
        self.get_helper().set_n(len(xs))
        result = xs if inplace else xs[:]
        self.aux = result[:]
        self._sort(result, 0, len(result))
        return result

    def _sort(self, xs: List[C], _from: int, _to: int) -> None:
        lo = _from
        hi = _to
        if hi <= lo + 1:
            return
        mid = _from + (_to - _from) // 2
        self._sort(xs, lo, mid)
        self._sort(xs, mid, hi)
        for i in range(_from, _to):
            self.aux[i] = xs[i]
        self.merge(self.aux, xs, lo, mid, hi)

    def merge(self, aux: List[C], xs: List[C], lo: int, mid: int, hi: int) -> None:
        i = lo
        j = mid
        for k in range(lo, hi):
            if i >= mid:
                xs[k] = aux[j]
                j += 1
            elif j >= hi:
                xs[k] = aux[i]
                i += 1
            elif self.helper.less(aux[j], aux[i]):
                xs[k] = aux[j]
                j += 1
            else:
                xs[k] = aux[i]
                i += 1

    def get_helper(self) -> Helper[C]:
        return self.helper
